### ISSUE TITLE
[MNG-6586] Allow proxy activation based on profile

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8664]
**[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl)** opened **[MNG-6586](https://issues.apache.org/jira/browse/MNG-6586?redirect=false)** and commented

Hello,

In setting reference[1] it is written:

> |`proxies/[proxy](https://maven.apache.org/ref/3.6.0/maven-settings/settings.html#class_proxy)*`|`List<Proxy>`|**(Many)** Configuration for different proxy profiles. Multiple proxy profiles might come in handy for anyone working from a notebook or other mobile platform, to enable easy switching of entire proxy configurations by simply specifying the profile id, again either from the command line or from the defaults section below.|

So it should be possible to enable/disable proxy based on a profile. However, there is no instructions of how to do that.

Adding \<active>${property}\</active> automatically disables the entry, even if property is set to true, I verify this using help:effective-settings. I could not find any property that can make any difference.

I see many people have tried and failed, and based on the above documentation it seems like either missing documentation or an actual bug.

Please allow that, as multiple setting.xml are not supported and enable/disable proxy is required for a single CI settings.xml.

Thanks!

 

[1] https://maven.apache.org/ref/3.6.0/maven-settings/settings.html#settings

---

**Affects:** 3.6.0

**Issue Links:**
- [MNG-5698](https://issues.apache.org/jira/browse/MNG-5698) Declare proxies and mirrors in profiles

- [MNG-6401](https://issues.apache.org/jira/browse/MNG-6401) Support proxy port interpolation in settings.xml

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think the wording is just bad. These profiles have likely nothing to do with Maven Profile. Maybe the word configuration would be better suited.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

As far as I can see profiles cannot contain proxy config. I will change this piece of documentation.

**Comment by jira-importer** [comment]
**[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl)** commented

This is not the answer I hoped to receive... :(

 

The expected configuration based on the text would be:

```java
  <proxies>
    <proxy>
      <id>id1</id>
      <active>${proxy.active}</active>
    </proxy>
  </proxies>
  <profiles>
    <profile>
      <id>enable-proxy</id>
      <properties>
        <proxy.active>true</proxy.active>
      </properties>
    </profile>
  </profiles>
```

 

Should I open a new bug?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will double-check before changing the docs.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This is by design and basically related to MNG-6401. This value cannot be interpolated and cannot be deferred. It is immediately coerced to boolean. Thus, always false becaue interpolation must happen before type system kicks in. [Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) solved this in [this branch](https://github.com/apache/maven/commit/0fb40ce55bf4de7ec98a55bd5fb9ff3a92298a19). Please try. We maybe need to reconsider MNG-6401.

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), WDYT?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

We need at least chage the documentation.

**Comment by jira-importer** [comment]
**[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl)** commented

Thanks for the hint, I missed it when searched for opened bugs. Yes, it seems like long standing issue, I found MNG-2465 and its references.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

In the end I would like a redesign of the `settings.properties`. Just wondering: if you need to add something to commandline to activate the proxy, how about activating a different settings with `\-\-settings,-s <arg>` or `\-\-global-settings,-gs <arg>` ?

**Comment by jira-importer** [comment]
**[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl)** commented

> how about activating a different settings with `--settings,-s <arg>` or `--global-settings,-gs <arg>` ?

 

Had maven would have supported multiple settings, this would have been an option... I do not mind to specify -gs gs.xml -s s.xml -s proxy.xml, however, as far as I know this is not possible.

 

The issue is that proxy (and mirrors) are environmental setting, and are set on top of the build settings.

 

If we reopen the entire subject, it is unclear why the system setting proxy (via http*_proxy_* is not honored by maven... this would have made usage so much easier... as you can notice in MNG-6401, using environment variable is preferred method.

**Comment by jira-importer** [comment]
**[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl)** commented

I just want to add how I manage this as of now... as I cannot include multiple settings, and it is difficult to merge XMLs, I use the fact that everything in active that is not 'true' is 'false'... so I have:

 
\<proxies>
    \<proxy>
      \<id>id1\</id>
      \<active>PROXY_ID1_ENABLE\</active>
    \</proxy>
  \</proxies>
 

And use sed 's/PROXY_ID1_ENABLE/true/g' mysettings.xml > mysettings1.xml when I need to enable the proxy and use this settings.

This is far from being ideal.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), I don't think we can fix this within weeks. I'd change the documentation for now. Is that acceptable?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Alon Bar-Lev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alonbl) you can have one global `settings.xml` (by default in apache-maven-3.x.x/conf) and one usr `settings.xml` (by default in ${user.home}/.m2) and these can be merged.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) agreed, this one is hard to fix. We should improve https://maven.apache.org/ref/3.6.0/maven-settings-builder/ just like https://maven.apache.org/ref/3.6.0/maven-model-builder/ to understand how it all works.
